### DISCUSSION TITLE
How to resolve this AirflowProviderDeprecationWarning?

### DISCUSSION THREAD

**Discussion by shreyateeza** [apache/airflow#41326] [Q&A]
ardvarc-airflow-worker-5db894c465-nr7l6
*** Found local files:
***   * /opt/airflow/logs/dag_id=spark_sql_operator_dag_2/run_id=manual__2024-08-08T08:47:20.718907+00:00/task_id=create_table_job/attempt=1.log
[2024-08-08, 08:47:32 UTC] {local_task_job_runner.py:120} ▶ Pre task execution logs
[2024-08-08, 08:47:32 UTC] {warnings.py:110} WARNING - /home/airflow/.local/lib/python3.12/site-packages/airflow/providers/apache/spark/operators/spark_sql.py:119: **AirflowProviderDeprecationWarning: Call to deprecated function (or staticmethod) _sql. (`_sql` is deprecated and will be removed in the future. Please use `sql` instead.**)
  sql=self._sql,
[2024-08-08, 08:47:33 UTC] {taskinstance.py:441} ▶ Post task execution logs

**Comment by omkar-foss** [comment]
@shreyateeza in your DAG, you'll need to check for places in code where `_sql` is being used with a Spark SQL operator. Most likely it will be used as a template variable, and that'd be triggering these deprecation warnings.

You'd simply need to replace those `_sql` with `sql` and the deprecation warnings should get resolved.

Update: For your reference, this deprecation warning was added in [this PR](https://github.com/apache/airflow/pull/38045/files#diff-756989f87fd48995dd96121179828a163f35c83f4dc7e500aad4b9579cb50147R96-R104).

**Reply by shreyateeza** [reply]
Hi @omkar-foss , in all the places, I have used 'sql' only. Nowhere is the '_sql' is used. Still I am getting this warning.

**Reply by shreyateeza** [reply]
Hi @omkar-foss , this is the piece of code I am using-

```
from airflow import DAG
from airflow.providers.apache.spark.operators.spark_sql import SparkSqlOperator
from airflow.utils.dates import days_ago
from datetime import timedelta
from pyspark import SparkConf


default_args = {
    'retries': 0,
    'retry_delay': timedelta(minutes=5),
}

spark_conf = SparkConf()
spark_conf.set("spark.sql.legacy.createHiveTableByDefault", "false")

sqlList = ['drop table if exists prod_ardvarc.mits_npe_stats', 'create table prod_ardvarc.mits_npe_stats as select load_timestamp, count(*) row_count from prod_ardvarc.tbl_mits_npe group by load_timestamp']

with DAG(
    dag_id = 'spark_sql_operator_dag_2',
    description = 'Running a SQL query',
    default_args = default_args,
    start_date = days_ago(1),
    schedule_interval = '@once',
    tags = ['python', 'spark', 'sql']
) as dag:
    
    dropTableIfExists = SparkSqlOperator(
        task_id="drop_table_job",
        sql=sqlList[0], 
        master="local" 
    )

    createTableMits_npe_stats = SparkSqlOperator(
        task_id='create_table_job',
        conf=spark_conf,
        sql=sqlList[1],
        master="local" 
    )

dropTableIfExists >> createTableMits_npe_stats

```

**Reply by phi-friday** [reply]
> Hi @omkar-foss , in all the places, I have used 'sql' only. Nowhere is the '_sql' is used. Still I am getting this warning.

Yes. It is being used internally by the provider. The warning will continue to occur until it is fixed with a PR.
https://github.com/apache/airflow/blob/06d6ecadffe7734e8758cf0067c73fd81d73c355/airflow/providers/apache/spark/operators/spark_sql.py#L119

**Reply by phi-friday** [reply]
fixed in #41358

**Reply by omkar-foss** [reply]
>Hi @omkar-foss , in all the places, I have used 'sql' only. Nowhere is the '_sql' is used. Still I am getting this warning.

There was also a line in Airflow code where `_sql` was being used, promptly fixed in #41358 (thanks @phi-friday).

@shreyateeza since your DAG doesn't have any direct `_sql` usage, the fix #41358 should completely resolve your deprecation warnings. It is now in the `main` branch, please check it out if possible.

**Reply by shreyateeza** [reply]
Thanks @phi-friday and @omkar-foss
